docs/how-does-version-resolution-work-in-maven-and-gradle.md

```diff
@@ -0,0 +1,124 @@
+
+So why do we need to pick a single version? The reason is that Java's default
+class loader only loads one version of each fully-qualified class name at
+runtime. So, if you try to put multiple versions on the classpath, you would not
+get both versions at runtime. Java build systems handle this issue upfront by
+picking a single version for each library before the classpath is even
+constructed.
+
+Now let's look at the version resolution algorithms used by Gradle and
+Maven. Let's consider Gradle first because it is the simpler one to
+describe. Simply put, Gradle chooses the highest version encountered in your
+dependency graph. For the example above, Gradle would pick A:10.1.
+
+<img src="assets/images/ddc-resolution-04.png">
+
+Gradle applies the higher-version rule even if it overrides one of your own
+direct dependency declarations:
+
+<img src="assets/images/ddc-resolution-05.png">
+
+Next let's consider Maven. Maven chooses the version that is closest to the root
+of the dependency graph, and if there is a tie, then it chooses the version it
+encounters first. Another way to describe the algorithm is that Maven chooses
+the first version encountered when traversing the dependency graph following a
+breadth-first traversal. Looking at the same example we used first for Gradle:
+
+<img src="assets/images/ddc-resolution-06.png">
+
+As you can see if you compare the resolved graphs, Maven makes the opposite
+choice. Let's look at the second example, where the root has both a direct and
+indirect dependency on the same library:
+
+<img src="assets/images/ddc-resolution-07.png">
+
+In this case, Maven chooses A:10.0 because it is only 1 hop away from the root
+instead of 2 hops (A:10.1). In this case, Maven also makes the opposite choice
+of Gradle.
+
+## Interaction between Maven and Gradle
+
+Every time a library is built, its build system performs version resolution for
+its entire dependency graph. Since the libraries in that dependency graph also
+perform their own version resolution, the versions selected can be different,
+especially when a consumer uses a different build system than its dependencies.
+Sometimes, the version selected by the consumer can be incompatible with the
+version selected by the dependency when the dependency is built by itself.
+
+Let's look at an example, starting with Gradle example 2, where library D
+depends on A:10.0 and C:30.0 (and thus indirectly A:10.1). Let's also say that C
+depends on a feature added into 10.1. This means that if 10.0 is selected as the
+resolved version, then C will fail at runtime. From library D's perspective,
+this is fine, since Gradle chooses version 10.1.
+
+Suppose we add a new library E which uses Maven as its build system, and which
+adds D as a dependency. When E is built, Maven resolves the whole dependency
+graph, including D's subgraph, even though Gradle resolved the subgraph of D for
+itself when D was originally built. When Maven performs its version resolution,
```

Comment:
"itself" is unclear here. Does this refer to Gradle or D?

Reply:
Removing "itself", it's not necessary